### PR TITLE
Fix assertion 'bytes_copied > 0 || !string_len' in JSON.stringify()

### PR DESCRIPTION
**Pull request by rtakacs**
Modified some places to support unicode characters.
Also renamed some variables to be more clearer.

### CODE REVIEW

**Comment by rtakacs** [comment]
Solution for #406

File: jerry-core/ecma/builtin-objects/ecma-builtin-json.cpp [hunk 1]
```diff
@@ -1117,11 +1117,11 @@ ecma_builtin_json_stringify (ecma_value_t this_arg __attr_unused___, /**< 'this'
     {
       ecma_length_t string_len = ecma_string_get_length (space_str_p);
```
Comment (zherczeg) [inline jerry-core/ecma/builtin-objects/ecma-builtin-json.cpp]: Are you sure this is length not size?

File: jerry-core/ecma/builtin-objects/ecma-builtin-json.cpp [hunk 2]
```diff
@@ -1198,23 +1198,24 @@ ecma_builtin_json_quote (ecma_string_t *string_p) /**< string that should be quo
   ecma_string_t *product_str_p = ecma_copy_or_ref_ecma_string (quote_str_p);
   ecma_string_t *tmp_str_p;
 
-  ecma_length_t string_len = ecma_string_get_length (string_p);
+  ecma_length_t string_size = ecma_string_get_size (string_p);
```
Comment (zherczeg) [inline jerry-core/ecma/builtin-objects/ecma-builtin-json.cpp]: I see you changed length to size here. I think you should do this above as well.

File: jerry-core/ecma/builtin-objects/ecma-builtin-json.cpp [hunk 3]
```diff
@@ -1130,13 +1130,13 @@ ecma_builtin_json_stringify (ecma_value_t this_arg __attr_unused___, /**< 'this'
 
       for (uint32_t i = 0; i < 10; i++)
       {
-        space_buff[i] = zt_string_buff[i];
+        space_buff[i] = string_buff[i];
       }
```
Comment (zherczeg) [inline jerry-core/ecma/builtin-objects/ecma-builtin-json.cpp]: What happens if the string length is < 10 byte?

Comment (rtakacs) [inline jerry-core/ecma/builtin-objects/ecma-builtin-json.cpp]: There is a condition above
```
if (num_of_chars < 10)
{
  context_p.gap_str_p = ecma_copy_or_ref_ecma_string (space_str_p);
}
else
{
  ...
}
```
If the number of characters are bigger than 10, the first 10 characters will be used.
If the number of characters are less than 10, all the characters will be used.

File: jerry-core/ecma/builtin-objects/ecma-builtin-json.cpp (hunk 2 above)
Comment (rtakacs) [inline jerry-core/ecma/builtin-objects/ecma-builtin-json.cpp]: You are right. Fixed.

**Comment by zherczeg** [comment]
LGTM

File: jerry-core/ecma/builtin-objects/ecma-builtin-json.cpp [hunk 4]
```diff
@@ -1115,28 +1115,27 @@ ecma_builtin_json_stringify (ecma_value_t this_arg __attr_unused___, /**< 'this'
...
-      {
-        space_buff[i] = zt_string_buff[i];
-      }
+      uint32_t space_buff_size = iter.buf_pos.offset;
+      /* Buffer for the first 10 characters. */
+      MEM_DEFINE_LOCAL_ARRAY (space_buff, space_buff_size, lit_utf8_byte_t);
+      memcpy (space_buff, string_buff, space_buff_size);
 
-      context_p.gap_str_p = ecma_new_ecma_string_from_utf8 ((lit_utf8_byte_t *) space_buff, 10);
+      context_p.gap_str_p = ecma_new_ecma_string_from_utf8 ((lit_utf8_byte_t *) space_buff, space_buff_size);
```
Comment (sand1k) [inline jerry-core/ecma/builtin-objects/ecma-builtin-json.cpp]: Do we really need space_buff?
As I see, we can just call `ecma_new_ecma_string_from_utf8 (string_buff, space_buff_size)`.

Comment (rtakacs) [inline jerry-core/ecma/builtin-objects/ecma-builtin-json.cpp]: hm... You are right, we don't need space_buff.
Updating.

File: jerry-core/ecma/builtin-objects/ecma-builtin-json.cpp [hunk 5]
```diff
@@ -1115,28 +1115,21 @@ ecma_builtin_json_stringify (ecma_value_t this_arg __attr_unused___, /**< 'this'
...
-      {
-        space_buff[i] = zt_string_buff[i];
-      }
+      context_p.gap_str_p = ecma_new_ecma_string_from_utf8 (string_buff, iter.buf_pos.offset);
 
-      context_p.gap_str_p = ecma_new_ecma_string_from_utf8 ((lit_utf8_byte_t *) space_buff, 10);
-
-      MEM_FINALIZE_LOCAL_ARRAY (space_buff);
-      MEM_FINALIZE_LOCAL_ARRAY (zt_string_buff);
+      MEM_FINALIZE_LOCAL_ARRAY (string_buff);
```
Comment (sand1k) [inline jerry-core/ecma/builtin-objects/ecma-builtin-json.cpp]: To extract a substirng, it is better to use:
`context_p.gap_str_p = ecma_string_substr (space_str_p, 0, 10)`.

Sorry, noticed just now, that this fragment could be simplified so much.

File: jerry-core/ecma/builtin-objects/ecma-builtin-json.cpp [hunk 6]
```diff
@@ -1115,28 +1115,21 @@ ecma_builtin_json_stringify (ecma_value_t this_arg __attr_unused___, /**< 'this'
...
-      /* Buffer for the first 10 characters. */
-      MEM_DEFINE_LOCAL_ARRAY (space_buff, 10, lit_utf8_byte_t);
+      lit_utf8_iterator_t iter = lit_utf8_iterator_create (string_buff, string_size);
+      lit_utf8_iterator_advance (&iter, 10);
 
-      for (uint32_t i = 0; i < 10; i++)
-      {
-        space_buff[i] = zt_string_buff[i];
-      }
+      context_p.gap_str_p = ecma_new_ecma_string_from_utf8 (string_buff, iter.buf_pos.offset);
```
Comment (sand1k) [inline jerry-core/ecma/builtin-objects/ecma-builtin-json.cpp]: And one more issue, `context_p` is not a pointer, could you please rename it to `context`?

Comment (rtakacs) [inline jerry-core/ecma/builtin-objects/ecma-builtin-json.cpp]: Indeed. Of course!

File: jerry-core/ecma/builtin-objects/ecma-builtin-json.cpp (hunk 5 above)
Comment (rtakacs) [inline jerry-core/ecma/builtin-objects/ecma-builtin-json.cpp]: nice! FIxed :)

**Comment by sand1k** [comment]
LGTM.

**Pull request merged**